### DISCUSSION TITLE
Using pgxpool and the connection goes away

### DISCUSSION THREAD

**Discussion by nchelluri** [jackc/pgx#2155] [Q&A] [answered]
Hello, I am using [pgxpool v5](https://pkg.go.dev/github.com/jackc/pgx/v5/pgxpool) as a connection pool. I am not sure what I am doing is the right way to do things (please let me know if you can see a better way), but I have created a health check method in my DB code and that calls `pool.Ping(ctx)` and if the call returns without error my code thinks the connection is healthy, otherwise it thinks the connection is unhealthy.

In my manual testing, I started a Pg server, connected a pool to it using `pgxpool.New()` (when creating a pool I call `pool.Ping()` and halt if an error occurs - I think this is reasonable for my use case) then I stopped the server and tried calling my health check method just to see what the return value looked like - I had expected the `pool.Ping()` call to return an error, but actually it never returns. Should it return? I think the call to `p.Acquire()` here doesn't complete: https://github.com/jackc/pgx/blob/v5.6.0/pgxpool/pool.go#L711

What should I be doing here? Questions I have are:

1. How do I test the health of the pool if not through `pool.Ping()`?
2. Will the pool recover if a connection goes bad?

Thanks for your help and I appreciate all of the work on this library.

**Comment by jackc** [comment, marked as answer]
>  I had expected the pool.Ping() call to return an error, but actually it never returns. 

If the connection is open as far as the client is concerned, then it may wait until the underlying net.Conn realizes it is dead. You may want to pass Ping a context with a deadline to handle broken connections.

**Comment by nchelluri** [comment]
Thanks, that is helpful.

And FWIW, I found out that the reason I ran into this issue at all is because I was using Pg via Docker, and I used `docker pause` to stop the container. I guess [I didn't realize it was different from `docker stop`](https://stackoverflow.com/questions/51466148/pause-vs-stop-in-docker). When I use `pause`, the context deadline/timeout seems to work well. And when I actually `docker stop` the container, the `pool.Ping()` call returns with a "failed to connect" error right away.
